olp-cpp-sdk-dataservice-read/src/VersionedLayerClientImpl.cpp

```diff
@@ -164,6 +167,164 @@ client::CancellableFuture<DataResponse> VersionedLayerClientImpl::GetData(
+                                  "PrefetchPartitions: getting versioned "
+                                  "partitions failed, key=%s",
+                                  request.CreateKey(layer_id).c_str());
+            return response.GetError();
+          }
+          PartitionDataHandleResult result;
+          std::transform(std::begin(response.GetResult().GetPartitions()),
+                         std::end(response.GetResult().GetPartitions()),
+                         std::inserter(result, result.end()),
+                         [&](const model::Partition& partition) {
+                           return std::make_pair(partition.GetPartition(),
+                                                 partition.GetDataHandle());
+                         });
+          return PartitionDataHandleResponse(result);
+        };
+
+        auto download = [=](std::string data_handle,
+                            client::CancellationContext inner_context) mutable {
+          if (data_handle.empty()) {
+            return BlobApi::DataResponse(
+                client::ApiError(client::ErrorCode::NotFound, "Not found"));
+          }
+          repository::DataCacheRepository data_cache_repository(
+              catalog, shared_settings->cache);
+          if (data_cache_repository.IsCached(layer_id, data_handle)) {
+            return BlobApi::DataResponse(nullptr);
+          }
+
+          repository::DataRepository repository(catalog, *shared_settings,
+                                                lookup_client);
+          // Fetch from online
+          return repository.GetVersionedData(layer_id,
+                                             DataRequest()
+                                                 .WithDataHandle(data_handle)
+                                                 .WithBillingTag(billing_tag),
+                                             version, inner_context);
+        };
+
+        // split prefetch ids to blocks by 100 items
+        auto size_left = request.GetPartitionIds().size();
+        std::vector<std::vector<std::string>> roots;
+        roots.reserve(size_left / kQueryPortion + 1);
+        auto start = 0u;
+        while (size_left > start) {
+          auto size =
+              std::min(static_cast<size_t>(kQueryPortion), size_left - start);
+          roots.emplace_back(std::vector<std::string>(
+              request.GetPartitionIds().begin() + start,
+              request.GetPartitionIds().begin() + start + size));
+          start += size;
+        }
+        auto append_result = [](ExtendedDataResponse response, std::string item,
+                                PrefetchPartitionsResult& prefetch_result) {
+          if (response.IsSuccessful()) {
+            prefetch_result.AddPartition(std::move(item));
+          }
+        };
+
+        context.ExecuteOrCancelled([&]() {
+          return PrefetchHelper::Prefetch<std::string, std::vector<std::string>,
```

Comment:
This PrefetchHelper::Prefetch() method became a monster. 3 template parameters and 9 input parameters. This needs some rework. I would propose to create a ticket in the backlog for this.